File: sdk/go/client_test.go
package aitrace

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestNewClient(t *testing.T) {
	client := NewClient("test-api-key")

	if client.apiKey != "test-api-key" {
		t.Errorf("Expected apiKey 'test-api-key', got '%s'", client.apiKey)
	}

	if client.baseURL != "https://api.aitrace.cc" {
		t.Errorf("Expected default baseURL, got '%s'", client.baseURL)
	}

	if client.Chat == nil {
		t.Error("Chat service should not be nil")
	}

	if client.Events == nil {
		t.Error("Events service should not be nil")
	}

	if client.Certs == nil {
		t.Error("Certs service should not be nil")
	}
}

func TestClientOptions(t *testing.T) {
	client := NewClient("test-api-key",
		WithBaseURL("https://custom.example.com"),
		WithUpstreamAPIKey("sk-upstream"),
		WithUpstreamBaseURL("https://upstream.example.com"),
		WithTimeout(30*time.Second),
	)

	if client.baseURL != "https://custom.example.com" {
		t.Errorf("Expected custom baseURL, got '%s'", client.baseURL)
	}

	if client.upstreamAPIKey != "sk-upstream" {
		t.Errorf("Expected upstream API key, got '%s'", client.upstreamAPIKey)
	}

	if client.upstreamBaseURL != "https://upstream.example.com" {
		t.Errorf("Expected upstream base URL, got '%s'", client.upstreamBaseURL)
	}

	if client.timeout != 30*time.Second {
		t.Errorf("Expected 30s timeout, got %v", client.timeout)
	}
}

func TestChatCreate(t *testing.T) {
	// Create mock server
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Verify request
		if r.Method != http.MethodPost {
			t.Errorf("Expected POST, got %s", r.Method)
		}

		if r.URL.Path != "/api/v1/chat/completions" {
			t.Errorf("Expected /api/v1/chat/completions, got %s", r.URL.Path)
		}

		if r.Header.Get("X-API-Key") != "test-api-key" {
			t.Errorf("Expected X-API-Key header")
		}

		// Return mock response
		resp := ChatResponse{
			ID:      "chatcmpl-123",
			Object:  "chat.completion",
			Created: time.Now().Unix(),
			Model:   "gpt-4",
			Choices: []Choice{
				{
					Index: 0,
					Message: Message{
						Role:    "assistant",
						Content: "Hello!",
					},
					FinishReason: "stop",
				},
			},
			Usage: Usage{
				PromptTokens:     10,
				CompletionTokens: 5,
				TotalTokens:      15,
			},
			TraceID: "trace-123",
		}

		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	// Create client with mock server
	client := NewClient("test-api-key", WithBaseURL(server.URL))

	// Make request
	resp, err := client.Chat.Create(context.Background(), ChatRequest{
		Model: "gpt-4",
		Messages: []Message{
			{Role: "user", Content: "Hello!"},
		},
	})

	if err != nil {
		t.Fatalf("Chat.Create failed: %v", err)
	}

	if resp.ID != "chatcmpl-123" {
		t.Errorf("Expected ID 'chatcmpl-123', got '%s'", resp.ID)
	}

	if resp.TraceID != "trace-123" {
		t.Errorf("Expected trace ID 'trace-123', got '%s'", resp.TraceID)
	}

	if len(resp.Choices) != 1 {
		t.Errorf("Expected 1 choice, got %d", len(resp.Choices))
	}

	if resp.Choices[0].Message.Content != "Hello!" {
		t.Errorf("Expected content 'Hello!', got '%s'", resp.Choices[0].Message.Content)
	}
}

func TestEventsIngest(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/events/ingest" {
			t.Errorf("Expected /api/v1/events/ingest, got %s", r.URL.Path)
		}

		resp := IngestResponse{
			Ingested: 2,
			EventIDs: []string{"event-1", "event-2"},
		}
		json.NewEncoder(w).Encode(resp)
	}))
	defer server.Close()

	client := NewClient("test-api-key", WithBaseURL(server.URL))

	events := []Event{
		{
			EventID:   "event-1",
			TraceID:   "trace-123",
			EventType: EventTypeInput,
			Payload:   map[string]interface{}{"prompt": "Hello"},
		},
		{
			EventID:   "event-2",
			TraceID:   "trace-123",
			EventType: EventTypeOutput,
			Payload:   map[string]interface{}{"content": "World"},
		},
	}

	resp, err := client.Events.Ingest(context.Background(), events)
	if err != nil {
		t.Fatalf("Events.Ingest failed: %v", err)
	}

	if resp.Ingested != 2 {
		t.Errorf("Expected 2 ingested, got %d", resp.Ingested)
	}
}

func TestCertsCommit(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/certs/commit" {
			t.Errorf("Expected /api/v1/certs/commit, got %s", r.URL.Path)
		}

		cert := Certificate{
			CertID:        "cert-123",
			TraceID:       "trace-123",
			RootHash:      "abc123",
			EventCount:    5,
			EvidenceLevel: EvidenceLevelL2,
			CreatedAt:     time.Now(),
		}
		json.NewEncoder(w).Encode(cert)
	}))
	defer server.Close()

	client := NewClient("test-api-key", WithBaseURL(server.URL))

	cert, err := client.Certs.Commit(context.Background(), "trace-123", EvidenceLevelL2)
	if err != nil {
		t.Fatalf("Certs.Commit failed: %v", err)
	}

	if cert.CertID != "cert-123" {
		t.Errorf("Expected cert ID 'cert-123', got '%s'", cert.CertID)
	}

	if cert.EvidenceLevel != EvidenceLevelL2 {
		t.Errorf("Expected evidence level L2, got '%s'", cert.EvidenceLevel)
	}
}

func TestCertsVerify(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/api/v1/certs/verify" {
			t.Errorf("Expected /api/v1/certs/verify, got %s", r.URL.Path)
		}

		result := VerificationResult{
			Valid: true,
			Checks: map[string]interface{}{
				"merkle_root":   true,
				"timestamp":     true,
				"event_hashes":  true,
			},
		}
		json.NewEncoder(w).Encode(result)
	}))
	defer server.Close()

	client := NewClient("test-api-key", WithBaseURL(server.URL))

	result, err := client.Certs.VerifyByCertID(context.Background(), "cert-123")
	if err != nil {
		t.Fatalf("Certs.Verify failed: %v", err)
	}

	if !result.Valid {
		t.Error("Expected valid certificate")
	}

	if result.Checks["merkle_root"] != true {
		t.Error("Expected merkle_root check to pass")
	}
}

func TestAPIError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(APIError{
			Code:    "invalid_request",
			Message: "Invalid trace ID",
		})
	}))
	defer server.Close()

	client := NewClient("test-api-key", WithBaseURL(server.URL))

	_, err := client.Certs.Commit(context.Background(), "", EvidenceLevelL1)
	if err == nil {
		t.Fatal("Expected error, got nil")
	}

	apiErr, ok := err.(*APIError)
	if !ok {
		t.Fatalf("Expected APIError, got %T", err)
	}

	if apiErr.Code != "invalid_request" {
		t.Errorf("Expected code 'invalid_request', got '%s'", apiErr.Code)
	}
}

func TestEventBuilder(t *testing.T) {
	event := NewEventBuilder("trace-123", EventTypeInput).
		WithSequence(1).
		AddPayloadField("prompt", "Hello").
		AddPayloadField("model_id", "gpt-4").
		Build()

	if event.TraceID != "trace-123" {
		t.Errorf("Expected trace ID 'trace-123', got '%s'", event.TraceID)
	}

	if event.EventType != EventTypeInput {
		t.Errorf("Expected event type '%s', got '%s'", EventTypeInput, event.EventType)
	}

	if event.Sequence != 1 {
		t.Errorf("Expected sequence 1, got %d", event.Sequence)
	}

	if event.Payload["prompt"] != "Hello" {
		t.Errorf("Expected prompt 'Hello', got '%v'", event.Payload["prompt"])
	}
}

func TestHelperFunctions(t *testing.T) {
	// Test InputEvent
	inputEvent := InputEvent("trace-123", "Hello", "gpt-4")
	if inputEvent.EventType != EventTypeInput {
		t.Errorf("Expected event type '%s', got '%s'", EventTypeInput, inputEvent.EventType)
	}

	// Test OutputEvent
	outputEvent := OutputEvent("trace-123", "World", 10)
	if outputEvent.EventType != EventTypeOutput {
		t.Errorf("Expected event type '%s', got '%s'", EventTypeOutput, outputEvent.EventType)
	}

	// Test ToolCallEvent
	toolEvent := ToolCallEvent("trace-123", "search", map[string]interface{}{"query": "test"})
	if toolEvent.EventType != EventTypeToolCall {
		t.Errorf("Expected event type '%s', got '%s'", EventTypeToolCall, toolEvent.EventType)
	}
}

func TestHashContent(t *testing.T) {
	hash1 := HashContent("test")
	hash2 := HashContent("test")
	hash3 := HashContent("different")

	if hash1 != hash2 {
		t.Error("Same content should produce same hash")
	}

	if hash1 == hash3 {
		t.Error("Different content should produce different hash")
	}

	if len(hash1) != 64 { // SHA256 produces 64 hex characters
		t.Errorf("Expected hash length 64, got %d", len(hash1))
	}
}

func TestIsValidEvidenceLevel(t *testing.T) {
	if !IsValidEvidenceLevel(EvidenceLevelL1) {
		t.Error("L1 should be valid")
	}
	if !IsValidEvidenceLevel(EvidenceLevelL2) {
		t.Error("L2 should be valid")
	}
	if !IsValidEvidenceLevel(EvidenceLevelL3) {
		t.Error("L3 should be valid")
	}
	if IsValidEvidenceLevel("L4") {
		t.Error("L4 should be invalid")
	}
}

func TestIsValidEventType(t *testing.T) {
	if !IsValidEventType(EventTypeInput) {
		t.Error("llm.input should be valid")
	}
	if !IsValidEventType(EventTypeOutput) {
		t.Error("llm.output should be valid")
	}
	if IsValidEventType("invalid.type") {
		t.Error("invalid.type should be invalid")
	}
}

File: sdk/javascript/src/__tests__/client.test.ts
import { describe, it, expect, beforeAll, afterAll, beforeEach } from 'vitest';
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';
import {
  AITraceClient,
  AITraceError,
  EventBuilder,
  MessageHelper,
  EvidenceLevel,
  EventType,
  ChatResponse,
  Certificate,
  VerificationResult,
  Event,
  IngestResponse,
} from '../index';

// Mock server setup
const mockServer = setupServer();

describe('AITraceClient', () => {
  const baseUrl = 'https://test.api.aitrace.cc';

  beforeAll(() => {
    mockServer.listen({ onUnhandledRequest: 'error' });
  });

  afterAll(() => {
    mockServer.close();
  });

  beforeEach(() => {
    mockServer.resetHandlers();
  });

  describe('constructor', () => {
    it('should create client with required options', () => {
      const client = new AITraceClient({
        apiKey: 'test-api-key',
        baseUrl,
      });

      expect(client).toBeInstanceOf(AITraceClient);
      expect(client.chat).toBeDefined();
      expect(client.events).toBeDefined();
      expect(client.certs).toBeDefined();
    });

    it('should throw error without apiKey', () => {
      expect(() => {
        new AITraceClient({ apiKey: '' });
      }).toThrow('apiKey is required');
    });

    it('should use default baseUrl', () => {
      const client = new AITraceClient({ apiKey: 'test-key' });
      expect(client).toBeInstanceOf(AITraceClient);
    });
  });

  describe('ChatService', () => {
    it('should create chat completion', async () => {
      const mockResponse: ChatResponse = {
        id: 'chatcmpl-123',
        object: 'chat.completion',
        created: 1704067200,
        model: 'gpt-4',
        choices: [
          {
            index: 0,
            message: { role: 'assistant', content: 'Hello!' },
            finish_reason: 'stop',
          },
        ],
        usage: { prompt_tokens: 10, completion_tokens: 5, total_tokens: 15 },
        trace_id: 'trace-123',
      };

      mockServer.use(
        http.post(`${baseUrl}/api/v1/chat/completions`, () => {
          return HttpResponse.json(mockResponse);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const response = await client.chat.create({
        model: 'gpt-4',
        messages: [{ role: 'user', content: 'Hi!' }],
      });

      expect(response.id).toBe('chatcmpl-123');
      expect(response.trace_id).toBe('trace-123');
      expect(response.choices[0].message.content).toBe('Hello!');
      expect(client.chat.lastTraceId).toBe('trace-123');
    });

    it('should create chat completion and commit certificate', async () => {
      const mockChatResponse: ChatResponse = {
        id: 'chatcmpl-123',
        object: 'chat.completion',
        created: 1704067200,
        model: 'gpt-4',
        choices: [{ index: 0, message: { role: 'assistant', content: 'Hello!' }, finish_reason: 'stop' }],
        trace_id: 'trace-123',
      };

      const mockCert: Certificate = {
        cert_id: 'cert-123',
        trace_id: 'trace-123',
        root_hash: 'abc123',
        event_count: 2,
        evidence_level: 'L2',
        created_at: '2024-01-01T00:00:00Z',
      };

      mockServer.use(
        http.post(`${baseUrl}/api/v1/chat/completions`, () => {
          return HttpResponse.json(mockChatResponse);
        }),
        http.post(`${baseUrl}/api/v1/certs/commit`, () => {
          return HttpResponse.json(mockCert);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const result = await client.chat.createAndCommit(
        { model: 'gpt-4', messages: [{ role: 'user', content: 'Hi!' }] },
        EvidenceLevel.L2
      );

      expect(result.chatResponse.trace_id).toBe('trace-123');
      expect(result.certificate.cert_id).toBe('cert-123');
    });
  });

  describe('EventsService', () => {
    it('should ingest events', async () => {
      const mockResponse: IngestResponse = {
        ingested: 2,
        event_ids: ['event-1', 'event-2'],
      };

      mockServer.use(
        http.post(`${baseUrl}/api/v1/events/ingest`, () => {
          return HttpResponse.json(mockResponse);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const events = [
        EventBuilder.input('trace-123', 'Hello', 'gpt-4'),
        EventBuilder.output('trace-123', 'Hi there!', 10),
      ];

      const response = await client.events.ingest(events);

      expect(response.ingested).toBe(2);
      expect(response.event_ids).toHaveLength(2);
    });

    it('should search events', async () => {
      mockServer.use(
        http.get(`${baseUrl}/api/v1/events/search`, ({ request }) => {
          const url = new URL(request.url);
          expect(url.searchParams.get('trace_id')).toBe('trace-123');
          return HttpResponse.json({
            events: [],
            total: 0,
            page: 1,
            page_size: 20,
            total_pages: 0,
          });
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const response = await client.events.search({ traceId: 'trace-123' });

      expect(response.events).toEqual([]);
    });

    it('should get event by ID', async () => {
      const mockEvent: Event = {
        event_id: 'event-123',
        trace_id: 'trace-123',
        event_type: 'llm.input',
        payload: { prompt: 'Hello' },
      };

      mockServer.use(
        http.get(`${baseUrl}/api/v1/events/event-123`, () => {
          return HttpResponse.json(mockEvent);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const event = await client.events.get('event-123');

      expect(event.event_id).toBe('event-123');
    });
  });

  describe('CertsService', () => {
    it('should commit certificate', async () => {
      const mockCert: Certificate = {
        cert_id: 'cert-123',
        trace_id: 'trace-123',
        root_hash: 'abc123',
        event_count: 5,
        evidence_level: 'L2',
        created_at: '2024-01-01T00:00:00Z',
      };

      mockServer.use(
        http.post(`${baseUrl}/api/v1/certs/commit`, () => {
          return HttpResponse.json(mockCert);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const cert = await client.certs.commit('trace-123', EvidenceLevel.L2);

      expect(cert.cert_id).toBe('cert-123');
      expect(cert.evidence_level).toBe('L2');
    });

    it('should verify certificate by ID', async () => {
      const mockResult: VerificationResult = {
        valid: true,
        checks: {
          merkle_root: true,
          timestamp: true,
          event_hashes: true,
          causal_chain: true,
        },
      };

      mockServer.use(
        http.post(`${baseUrl}/api/v1/certs/verify`, () => {
          return HttpResponse.json(mockResult);
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const result = await client.certs.verifyByCertId('cert-123');

      expect(result.valid).toBe(true);
      expect(result.checks.merkle_root).toBe(true);
    });

    it('should generate proof', async () => {
      mockServer.use(
        http.post(`${baseUrl}/api/v1/certs/cert-123/prove`, () => {
          return HttpResponse.json({
            cert_id: 'cert-123',
            root_hash: 'abc123',
            disclosed_events: [],
            merkle_proofs: [],
            metadata: {},
          });
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });
      const proof = await client.certs.prove('cert-123', { discloseEvents: [0, 2] });

      expect(proof.cert_id).toBe('cert-123');
    });

    it('should use convenience methods for evidence levels', async () => {
      const createMockHandler = (expectedLevel: string) =>
        http.post(`${baseUrl}/api/v1/certs/commit`, async ({ request }) => {
          const body = await request.json() as { evidence_level: string };
          expect(body.evidence_level).toBe(expectedLevel);
          return HttpResponse.json({
            cert_id: 'cert-123',
            trace_id: 'trace-123',
            root_hash: 'abc',
            event_count: 1,
            evidence_level: expectedLevel,
            created_at: '2024-01-01T00:00:00Z',
          });
        });

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });

      mockServer.use(createMockHandler('L1'));
      await client.certs.commitL1('trace-123');

      mockServer.use(createMockHandler('L2'));
      await client.certs.commitL2('trace-123');

      mockServer.use(createMockHandler('L3'));
      await client.certs.commitL3('trace-123');
    });
  });

  describe('Error Handling', () => {
    it('should throw AITraceError on API error', async () => {
      mockServer.use(
        http.post(`${baseUrl}/api/v1/certs/commit`, () => {
          return HttpResponse.json(
            { code: 'invalid_request', message: 'Invalid trace ID' },
            { status: 400 }
          );
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });

      await expect(client.certs.commit('', EvidenceLevel.L1)).rejects.toThrow(AITraceError);

      try {
        await client.certs.commit('', EvidenceLevel.L1);
      } catch (error) {
        expect(error).toBeInstanceOf(AITraceError);
        const aiError = error as AITraceError;
        expect(aiError.code).toBe('invalid_request');
        expect(aiError.statusCode).toBe(400);
        expect(aiError.isClientError()).toBe(true);
        expect(aiError.isServerError()).toBe(false);
      }
    });

    it('should handle server errors', async () => {
      mockServer.use(
        http.get(`${baseUrl}/api/v1/events/test`, () => {
          return HttpResponse.json(
            { code: 'internal_error', message: 'Server error' },
            { status: 500 }
          );
        })
      );

      const client = new AITraceClient({ apiKey: 'test-key', baseUrl });

      try {
        await client.events.get('test');
      } catch (error) {
        expect(error).toBeInstanceOf(AITraceError);
        const aiError = error as AITraceError;
        expect(aiError.isServerError()).toBe(true);
      }
    });
  });
});

describe('EventBuilder', () => {
  it('should build event with builder pattern', () => {
    const event = new EventBuilder()
      .traceId('trace-123')
      .eventType(EventType.INPUT)
      .sequence(1)
      .addPayload('prompt', 'Hello')
      .addPayload('model_id', 'gpt-4')
      .build();

    expect(event.trace_id).toBe('trace-123');
    expect(event.event_type).toBe('llm.input');
    expect(event.sequence).toBe(1);
    expect(event.payload.prompt).toBe('Hello');
    expect(event.payload.model_id).toBe('gpt-4');
  });

  it('should throw error without required fields', () => {
    expect(() => new EventBuilder().build()).toThrow('trace_id is required');
    expect(() => new EventBuilder().traceId('t').build()).toThrow('event_type is required');
  });

  it('should create input event', () => {
    const event = EventBuilder.input('trace-123', 'Hello', 'gpt-4');
    expect(event.event_type).toBe('llm.input');
    expect(event.payload.prompt).toBe('Hello');
    expect(event.payload.model_id).toBe('gpt-4');
  });

  it('should create output event', () => {
    const event = EventBuilder.output('trace-123', 'Hi there!', 10);
    expect(event.event_type).toBe('llm.output');
    expect(event.payload.content).toBe('Hi there!');
    expect(event.payload.tokens).toBe(10);
  });

  it('should create tool call event', () => {
    const event = EventBuilder.toolCall('trace-123', 'search', { query: 'test' });
    expect(event.event_type).toBe('llm.tool_call');
    expect(event.payload.tool_name).toBe('search');
    expect(event.payload.arguments).toEqual({ query: 'test' });
  });

  it('should create tool result event', () => {
    const event = EventBuilder.toolResult('trace-123', 'search', { results: [] });
    expect(event.event_type).toBe('llm.tool_result');
    expect(event.payload.result).toEqual({ results: [] });
  });

  it('should create error event', () => {
    const event = EventBuilder.error('trace-123', 'ERR001', 'Something went wrong');
    expect(event.event_type).toBe('llm.error');
    expect(event.payload.error_code).toBe('ERR001');
    expect(event.payload.message).toBe('Something went wrong');
  });
});

describe('MessageHelper', () => {
  it('should create user message', () => {
    const msg = MessageHelper.user('Hello');
    expect(msg.role).toBe('user');
    expect(msg.content).toBe('Hello');
  });

  it('should create assistant message', () => {
    const msg = MessageHelper.assistant('Hi there!');
    expect(msg.role).toBe('assistant');
    expect(msg.content).toBe('Hi there!');
  });

  it('should create system message', () => {
    const msg = MessageHelper.system('You are helpful.');
    expect(msg.role).toBe('system');
    expect(msg.content).toBe('You are helpful.');
  });
});

describe('Constants', () => {
  it('should have correct evidence levels', () => {
    expect(EvidenceLevel.L1).toBe('L1');
    expect(EvidenceLevel.L2).toBe('L2');
    expect(EvidenceLevel.L3).toBe('L3');
  });

  it('should have correct event types', () => {
    expect(EventType.INPUT).toBe('llm.input');
    expect(EventType.OUTPUT).toBe('llm.output');
    expect(EventType.CHUNK).toBe('llm.chunk');
    expect(EventType.TOOL_CALL).toBe('llm.tool_call');
    expect(EventType.TOOL_RESULT).toBe('llm.tool_result');
    expect(EventType.ERROR).toBe('llm.error');
  });
});
